Restructuring Socket/TCP conceptual docs PR (#32022)
dotnet/docs: docs/fundamentals/networking/sockets/tcp-classes.md

---

> title: Use TcpClient and TcpListener in .NET

Reviewer: ```suggestion
title: Use TcpClient and TcpListener
```
We already have a ".NET" suffix added.

description: Learn how to use the TcpClient class to create a socket to request and receive data using TCP in .NET.
author: IEvangelist
ms.author: dapine
ms.date: 08/24/2022
helpviewer_keywords: 
  - "protocols, TCP"
  - "network resources, TCP"
  - "sending data, TCP"
  - "TCP"
  - "TcpClient class, about TcpClient class"
  - "application protocols, TCP"
  - "receiving data, TCP"
  - "TcpListener class, about TcpListener class"
  - "Socket class, about Socket class"
  - "data requests, TCP"
  - "requesting data from Internet, TCP"
  - "Internet, TCP"
---

> # TCP Overview

Reviewer: ```suggestion
# TCP overview
```

We follow sentence casing for headings

> [!IMPORTANT]
> <xref:System.Net.Sockets.Socket> class is highly recommended for advanced users, instead of `TcpClient` and `TcpListener`.

> To use the Transmission Control Protocol (TCP) services in .NET, you have two options, one of them is using the <xref:System.Net.Sockets.Socket> which is highly recommended, or use the <xref:System.Net.Sockets.TcpClient> and <xref:System.Net.Sockets.TcpListener> classes. <xref:System.Net.Sockets.TcpClient> and <xref:System.Net.Sockets.TcpListener> are built on top of the <xref:System.Net.Sockets.Socket?displayProperty=nameWithType> class and take care of the details of transferring data and main purpose of these classes is ease of use.

Reviewer 2: ```suggestion
To work with Transmission Control Protocol (TCP), you have two options: either use <xref:System.Net.Sockets.Socket> for maximum control and performance, or use the <xref:System.Net.Sockets.TcpClient> and <xref:System.Net.Sockets.TcpListener> helper classes. <xref:System.Net.Sockets.TcpClient> and <xref:System.Net.Sockets.TcpListener> are built on top of the <xref:System.Net.Sockets.Socket?displayProperty=nameWithType> class and take care of the details of transferring data for ease of use.
```

The protocol classes use the underlying `Socket` class to provide simple access to network services without the overhead of maintaining state information or knowing the details of setting up protocol-specific sockets. To use asynchronous `Socket` methods, you can use the asynchronous methods supplied by the <xref:System.Net.Sockets.NetworkStream> class. To access features of the `Socket` class not exposed by the protocol classes, you must use the `Socket` class.

`TcpClient` and `TcpListener` represent the network using the `NetworkStream` class. You use the <xref:System.Net.Sockets.TcpClient.GetStream%2A> method to return the network stream, and then call the stream's <xref:System.Net.Sockets.NetworkStream.ReadAsync%2A?displayProperty=nameWithType> and <xref:System.Net.Sockets.NetworkStream.WriteAsync%2A?displayProperty=nameWithType> methods. The `NetworkStream` does not own the protocol classes' underlying socket, so closing it does not affect the socket.

> ## Use TcpClient and TcpListener

Reviewer: ```suggestion
## Use `TcpClient` and `TcpListener`
```
